Dqlite: Fix TLS handshake errors

roosterfish opened the pull request:
See https://github.com/canonical/go-dqlite/issues/398 for reference.
Closes https://github.com/canonical/go-dqlite/issues/398.

Microcluster's custom dial function passed into `go-dqlite` might cause TLS handshake errors on the target members in case the context passed into the dial function is cancelled early on. That might be the case when a leader was already found, so all the other parallel connection attempts to the other members are cancelled.
Whilst this is not a functional issue we should have extra handling to prevent seeing those error messages in the Microcluster daemon logs.

Furthermore the passed context's deadline is now used properly. Setting it whilst using the same context the deadline is coming from doesn't bring much value.
In addition a potential leak of the request's response body is fixed.

roosterfish committed:
internal/db: Only use outer context for TLS dial

Setting both the timeout (from the outer context's deadline) and the context for the dial is redundant.
Instead only use the outer context for the TLS dial.

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

roosterfish committed:
internal/db: Update debug message

We use lowercase dqlite in all other messages.

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

roosterfish committed:
internal/db: Don't leave the request body open in case of error

Add the response body to the reverter.
Also closing the body should succeed. If not it can be considered an error.

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

tomponline commented:
If deadline is zero, does WithDeadline never cancel or immediately cancel?

roosterfish committed:
internal/daemon: Add logfilter

This allows filtering out the TLS handshake error messages which might appear in case go-dqlite
cancels the connection attempts prematuerely in case the leader was already found.

It doesn't hide an error but instead suppresses errors which are considered to be not desired.

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

roosterfish committed:
internal/daemon: Set a custom logger for the underlying HTTP sever

This allows injecting our filter logger.

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

roosterfish commented:
> If deadline is zero, does WithDeadline never cancel or immediately cancel?

I got rid of this part in the changes. But I guess it will never cancel as deadline is zero. So it's the same as setting no deadline.

roosterfish commented:
@markylaing let's talk this through during our 1:1 next week.

markylaing commented on internal/daemon/logfilter.go:
```diff
@@ -0,0 +1,43 @@
+
+// logBlacklist represents log texts whose presence will cause log messages to be filtered out.
+var logBlacklist = []string{
+	"http: TLS handshake error from",
```
I think we should do a quick review of the errors that are logged by the go stdlib http server to ensure we aren't going to prevent logging other unexpected errors with this prefix. We might need to be a little more restrictive in the blacklist as LXD does with the regexp filtering

roosterfish replied:
As discussed earlier I'll validate how feasible it is to use the list of known cluster members so we only filter out error messages from our own members, but not from external clients talking to the API.

roosterfish replied:
Pushed the necessary changes.

I found that the prefix/newline stripping we are using in LXD's equivalent to this isn't required so I didn't include it. Also I have simplified the regex for the exact error message we are facing in this case.
Let me know what you think.

markylaing commented on internal/daemon/daemon.go:
```diff
@@ -495,9 +496,19 @@ func (d *Daemon) initServer(resources ...rest.Resources) *http.Server {
 	return &http.Server{
 		Handler:     mux,
 		ConnContext: request.SaveConnectionInContext,
+		ErrorLog:    log.New(newLogFilter(knownAddresses), "", 0),
```
Just for my understanding: When a new member joins or when a member is removed, is the http server restarted? E.g. will this list of known addresses always be correct?

roosterfish replied:
Good catch, no it won't get restarted. Looks like the log filter needs to query for the addresses on demand.

roosterfish replied:
Fixed. Also got rid of an obsolete extra loop to craft a slice of addresses which isn't anymore required. Before I was using strings to represent the addresses but have decided to use the `AddrPort` type directly for convenience.

roosterfish committed:
internal/daemon: Add temp truststore to test

This allows calling the daemon's internal remote's Addresses func which is now required as part of the log filtering
to gather a list of cluster member addresses.

Signed-off-by: Julian Pelizäus <julian.pelizaeus@canonical.com>

markylaing commented on internal/daemon/daemon.go:
```diff
@@ -498,6 +499,7 @@ func (d *Daemon) initServer(resources ...rest.Resources) *http.Server {
 	return &http.Server{
 		Handler:     mux,
 		ConnContext: request.SaveConnectionInContext,
+		ErrorLog:    log.New(newLogFilter(state.Remotes().Addresses), "", 0),
```
Last question - is calling `state.Remotes().Addresses()` expensive? I worry that if something causes the http server to spam logs we'll have trouble accessing the truststore.

IIRC this acquires a read lock on the truststore, which should be fine especially since a write lock is only acquired for infrequent trust store changes.

roosterfish replied:
You are right with the locks. I was also checking this code path before to be sure it's not too expensive. Inside `Addresses` there is a loop which creates the map and then releases the lock again.

roosterfish commented:
@markylaing when you are happy with the PR please leave it open and don't merge it yet. 
Whilst this fixes the error message we are seeing, I found we could maybe use a different method on `go-dqlite` to get to the leader which maybe won't cause those error messages on the target side. Testing this right now.

markylaing reviewed (approved):
LGTM thanks :partying_face:

roosterfish commented:
What I have mentioned in [this](https://github.com/canonical/microcluster/pull/470#issuecomment-3244611789) comment isn't a full alternative unfortunately. However we could think about using `go-dqlite`'s `FindLeader` instead of `Leader` to reuse an existing connection to the leader. This also reduces the likelihood of the error message we are trying to fix here, but doesn't remove it completely.

I'll add another PR for this and will merge this one now.
